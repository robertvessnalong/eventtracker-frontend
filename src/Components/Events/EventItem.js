import { useState, useContext } from 'react';
import {
  Card,
  Button,
  Container,
  Col,
  Form,
  Row,
  Alert,
} from 'react-bootstrap';
import UserContext from '../../Context/UserContext';
import EventFinderApi from '../../API/api';
import { AiFillHeart, AiOutlineHeart } from 'react-icons/ai';

const EventItem = ({ favorites, event }) => {
  const { user, setUser, loggedIn } = useContext(UserContext);
  const formInitalState = {
    eventId: event.id,
    eventName: event.title,
    userName: user.firstName,
    comment: '',
    uuid: user['uuid'],
  };
  const [formData, setFormData] = useState(formInitalState);
  const [comment, setIsCommenting] = useState(false);
  const [error, setError] = useState(false);
  const [errorMsg, setErrorMsg] = useState([]);

  const handleCommentSection = () => {
    setIsCommenting(!comment);
  };

  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData((data) => ({
      ...data,
      [name]: value,
    }));
  };

  // Handle Favorties
  const handleFavorites = async (e, event) => {
    const eventData = {
      type: 'event',
      eventId: event.id,
      eventName: event.title,
      eventDate: EventFinderApi.eventTimeConverter(event.datetime_utc),
      eventAddress: event.venue.address,
      eventExtendedAddress: event.venue.extended_address,
      eventUrl: event.url,
      uuid: user['uuid'],
    };
    const res = e.target.classList.contains('favorited')
      ? await EventFinderApi.removefavorite(eventData)
      : await EventFinderApi.favoriteEvent(eventData);
    if (res.error) {
      console.error(res.error);
    } else {
      const user = JSON.parse(localStorage.getItem('user'));
      const userData = await EventFinderApi.getUserInfo(user.user, user.token);
      setUser(userData);
    }
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    const res = await EventFinderApi.addComment(formData);
    if (res.error) {
      setError(true);
      setErrorMsg(res.error);
    } else if (res.comment) {
      const user = JSON.parse(localStorage.getItem('user'));
      const userData = await EventFinderApi.getUserInfo(user.user, user.token);
      setFormData(formInitalState);
      setIsCommenting(!comment);
      setUser(userData);
    }
  };

  return (
    <Card className='mt-4 mb-4 p-0'>
      <Card.Header as='h6'>
        <Row>
          <Col className='d-flex align-items-center' xs={10}>
            <small>{event.title}</small>
          </Col>
          <Col xs={2}>
            {loggedIn && (
              <>
                {EventFinderApi.favoriteCheck(favorites, event) ? (
                  <div
                    className='favorite-btn favorited'
                    onClick={(e) => handleFavorites(e, event)}
                  >
                    <AiFillHeart />
                  </div>
                ) : (
                  <div
                    className='favorite-btn'
                    onClick={(e) => handleFavorites(e, event)}
                  >
                    <AiOutlineHeart />
                  </div>
                )}
              </>
            )}
          </Col>
        </Row>
      </Card.Header>
      <Card.Body>
        <Row>
          <Col xs={12} md={6} className='text-start'>
            <div className='card-time'>
              <b>{EventFinderApi.eventTimeConverter(event.datetime_utc)}</b>
            </div>
            <address>
              {event.venue.address}
              <br />
              {event.venue.extended_address}
            </address>
          </Col>
          <Col
            xs={12}
            md={6}
            className='d-flex align-items-center justify-content-end'
          >
            <a href={event.url} rel='noreferrer' target='_blank'>
              <Button className='blue-btn' variant='primary'>
                Buy Tickets
              </Button>
            </a>
          </Col>
        </Row>
      </Card.Body>
      <Card.Footer className='text-muted text-start'>
        <button onClick={handleCommentSection}>Comment</button>
        <Col className={comment ? `Comment Active` : `Comment`} xs={12}>
          <Container>
            {error && <Alert variant='danger'>{errorMsg}</Alert>}
            <h6>Comment On Event</h6>
            <Form onSubmit={handleSubmit}>
              <Form.Control
                as='textarea'
                placeholder='Leave a comment here'
                name='comment'
                style={{ height: '150px' }}
                value={formData.comment}
                onChange={handleChange}
              />
              <button type='Submit'>Submit</button>
            </Form>
          </Container>
        </Col>
      </Card.Footer>
    </Card>
  );
};

export default EventItem;
