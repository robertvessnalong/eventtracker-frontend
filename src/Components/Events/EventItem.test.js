import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import EventItem from './EventItem';
import TestUser from '../../Testing/TestUser';
import TestEvent from '../../Testing/TestEvent';
import UserContext from '../../Context/UserContext';

describe('EventItem Component', () => {
  // Smoke Test
  test('renders without crashing', async () => {
    render(
      <UserContext.Provider value={{ user: TestUser }}>
        <EventItem event={TestEvent} favorites={TestUser.favorites} />
      </UserContext.Provider>
    );
    await waitFor(() => {
      const headerSpan = screen.getByText(
        /The Great Jack OLantern Blaze - Long Island/i
      );
      expect(headerSpan).toBeInTheDocument();
    });
  });

  //Snapshot Test
  test('matches snapshot', () => {
    const { asFragment } = render(
      <UserContext.Provider value={{ user: TestUser }}>
        <EventItem event={TestEvent} favorites={TestUser.favorites} />
      </UserContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  //Snapshot Logged In
  test('loggedIn snapshot', () => {
    const { asFragment } = render(
      <UserContext.Provider value={{ user: TestUser, loggedIn: true }}>
        <EventItem event={TestEvent} favorites={TestUser.favorites} />
      </UserContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('fireFavorite', async () => {
    const { container } = render(
      <UserContext.Provider value={{ user: TestUser, loggedIn: true }}>
        <EventItem event={TestEvent} favorites={TestUser.favorites} />
      </UserContext.Provider>
    );
    expect(container.getElementsByClassName('favorite-btn').length).toBe(1);
    fireEvent.click(container.querySelector('.favorite-btn'));
  });

  test('fireComment', async () => {
    const { container } = render(
      <UserContext.Provider value={{ user: TestUser, loggedIn: true }}>
        <EventItem event={TestEvent} favorites={TestUser.favorites} />
      </UserContext.Provider>
    );
    fireEvent.click(screen.getByText('Comment'));
    expect(container.getElementsByClassName('Active').length).toBe(1);
  });
});
