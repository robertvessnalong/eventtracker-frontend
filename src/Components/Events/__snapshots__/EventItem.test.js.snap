// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Events Component loggedIn snapshot 1`] = `
<DocumentFragment>
  <div
    class="mt-4 mb-4 p-0 card"
  >
    <h6
      class="card-header"
    >
      <div
        class="row"
      >
        <div
          class="d-flex align-items-center col-10"
        >
          <small>
            The Great Jack OLantern Blaze - Long Island
          </small>
        </div>
        <div
          class="col-2"
        >
          <div
            class="favorite-btn"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 1024 1024"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M923 283.6a260.04 260.04 0 0 0-56.9-82.8 264.4 264.4 0 0 0-84-55.5A265.34 265.34 0 0 0 679.7 125c-49.3 0-97.4 13.5-139.2 39-10 6.1-19.5 12.8-28.5 20.1-9-7.3-18.5-14-28.5-20.1-41.8-25.5-89.9-39-139.2-39-35.5 0-69.9 6.8-102.4 20.3-31.4 13-59.7 31.7-84 55.5a258.44 258.44 0 0 0-56.9 82.8c-13.9 32.3-21 66.6-21 101.9 0 33.3 6.8 68 20.3 103.3 11.3 29.5 27.5 60.1 48.2 91 32.8 48.9 77.9 99.9 133.9 151.6 92.8 85.7 184.7 144.9 188.6 147.3l23.7 15.2c10.5 6.7 24 6.7 34.5 0l23.7-15.2c3.9-2.5 95.7-61.6 188.6-147.3 56-51.7 101.1-102.7 133.9-151.6 20.7-30.9 37-61.5 48.2-91 13.5-35.3 20.3-70 20.3-103.3.1-35.3-7-69.6-20.9-101.9zM512 814.8S156 586.7 156 385.5C156 283.6 240.3 201 344.3 201c73.1 0 136.5 40.8 167.7 100.4C543.2 241.8 606.6 201 679.7 201c104 0 188.3 82.6 188.3 184.5 0 201.2-356 429.3-356 429.3z"
              />
            </svg>
          </div>
        </div>
      </div>
    </h6>
    <div
      class="card-body"
    >
      <div
        class="row"
      >
        <div
          class="text-start col-md-6 col-12"
        >
          <div
            class="card-time"
          >
            <b>
              Sun Oct 24 2021 - 7 AM
            </b>
          </div>
          <address>
            1303 Round Swamp Road
            <br />
            Old Bethpage, NY 11804
          </address>
        </div>
        <div
          class="d-flex align-items-center justify-content-end col-md-6 col-12"
        >
          <a
            href="https://seatgeek.com/the-great-jack-olantern-blaze-long-island-tickets/family/2021-10-24-3-30-am/5491086"
            rel="noreferrer"
            target="_blank"
          >
            <button
              class="blue-btn btn btn-primary"
              type="button"
            >
              Buy Tickets
            </button>
          </a>
        </div>
      </div>
    </div>
    <div
      class="text-muted text-start card-footer"
    >
      <button>
        Comment
      </button>
      <div
        class="Comment col-12"
      >
        <div
          class="container"
        >
          <h6>
            Comment On Event
          </h6>
          <form
            class=""
          >
            <textarea
              class="form-control"
              name="comment"
              placeholder="Leave a comment here"
              style="height: 150px;"
            />
            <button
              type="Submit"
            >
              Submit
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Events Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="mt-4 mb-4 p-0 card"
  >
    <h6
      class="card-header"
    >
      <div
        class="row"
      >
        <div
          class="d-flex align-items-center col-10"
        >
          <small>
            The Great Jack OLantern Blaze - Long Island
          </small>
        </div>
        <div
          class="col-2"
        />
      </div>
    </h6>
    <div
      class="card-body"
    >
      <div
        class="row"
      >
        <div
          class="text-start col-md-6 col-12"
        >
          <div
            class="card-time"
          >
            <b>
              Sun Oct 24 2021 - 7 AM
            </b>
          </div>
          <address>
            1303 Round Swamp Road
            <br />
            Old Bethpage, NY 11804
          </address>
        </div>
        <div
          class="d-flex align-items-center justify-content-end col-md-6 col-12"
        >
          <a
            href="https://seatgeek.com/the-great-jack-olantern-blaze-long-island-tickets/family/2021-10-24-3-30-am/5491086"
            rel="noreferrer"
            target="_blank"
          >
            <button
              class="blue-btn btn btn-primary"
              type="button"
            >
              Buy Tickets
            </button>
          </a>
        </div>
      </div>
    </div>
    <div
      class="text-muted text-start card-footer"
    >
      <button>
        Comment
      </button>
      <div
        class="Comment col-12"
      >
        <div
          class="container"
        >
          <h6>
            Comment On Event
          </h6>
          <form
            class=""
          >
            <textarea
              class="form-control"
              name="comment"
              placeholder="Leave a comment here"
              style="height: 150px;"
            />
            <button
              type="Submit"
            >
              Submit
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
